### PR TITLE
[ready for review] Pre-fill SubmissionInfo form of CfP with track and submission_type

### PR DESCRIPTION
**Pull request by Nakaner**
If the URL query string for the submission info form or the CfP landing page contains a track or submission_types field, they will be used to pre-fill the submission info form. If the values do not match any in the database, the fields will have their default as initial value.

Add slug fields to the SubmissionType and Track model in order to have a easy to query but still memorizeable (by humans) identifier because the i18n char fields are not designed to be queried.

Slugs for submission types and tracks can be choosen by organisers. The migration tries to set readable but unique values for existing entries in the database.

This allows to have two different CfP for different tracks of a conference outside of Pretalx and to point submitters to the "right" submission form for their track. In my case, a conference has about five "normal" and one academic track. The academic programme committee fears that their submitters don't select the right track due to the high number of thematic tracks for the normal part of the conference. (Sie sehen den Wald vor lauter Bäumen nicht) The behaviour of this feature is a bit like a anchor in an URL pointing to the right section of a longer web page.

I would like to have feedback on the implemention.

This feature is a feature known by experts at the moment due to a lack of a proper user documentation. Where shall I document it?

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [x] My change is listed in the CHANGELOG.rst if appropriate.

### CODE REVIEW

**Comment by Nakaner** [comment]
If I knew how to convert this pull request now to a normal pull request, I would do it. Unfortunately, the user interface lacks the button "mark ready for review" next to "This pull request is still a work in progress."

@rixx Shall I recreate this PR or can you review it and merge it manually?

**Comment by Nakaner** [comment]
I thought about serial vs. string for the slug field.

Pro string:

* It is more human readable in the URL.
* Uniqueness is only useful within an event (it is not enforced).
* A serial field makes porting an event into another instance with other existing events difficult.

Pro integer/serial field:

* performance

I am happy to change that if you ask me to do so.

File: doc/changelog.rst [hunk 1]
```diff
@@ -3,6 +3,7 @@
 Release Notes
 =============
 
+- :bug:`628` Prefill CfP form with track and submission type if they are available in the query string.
```
Comment (rixx) [inline doc/changelog.rst]: I think this is rather a `:feature:`

File: src/pretalx/cfp/views/event.py [hunk 2]
```diff
@@ -34,6 +36,19 @@ def get_context_data(self, **kwargs):
...
+        # Only used first value if key is used multiple times.
+        track = self.request.GET.get('track', None)
+        submission_type = self.request.GET.get('submission_type', None)
+        params = []
+        if track:
+            params.append(f'track={track}')
+        if submission_type:
+            params.append(f'submission_type={submission_type}')
+        if len(params) > 0:
+            context['submit_qs'] = f'?{"&".join(params)}'
```
Comment (rixx) [inline src/pretalx/cfp/views/event.py]: Letting `params` be a dict and using `urllib.urlencode(params)` would make sure all content is properly escaped. (Please test this with a track with special characters, such as umlauts.

File: src/pretalx/cfp/views/wizard.py [hunk 3]
```diff
@@ -127,6 +129,14 @@ def get_context_data(self, **kwargs):
         step_list.append({'phase': 'todo', 'label': _('Done!'), 'icon': 'check'})
         context['step_list'] = step_list
 
+        if step == 'info':
+            # use value of query string parameters track and submission_type as initial values for form if they are valid
+            for field, model_name in (('submission_type', SubmissionType), ('track', Track)):
+                request_value = self.request.GET.get(field)
+                if request_value:
+                    queryset = model_name.objects.filter(event=self.request.event, slug=request_value)
```
Comment (rixx) [inline src/pretalx/cfp/views/wizard.py]:
```
object = model_name.objects.filter(event=self.request.event, slug=request_value).first()
if object:
…
```

This avoids sending the ``COUNT`` to the database.

**Comment by Nakaner** [comment]
The commit I pushed a minute ago should address all your issues. I introduced a new mixin IdBasedMixin for the slug you described.

You reorganised the tests last night. Do you expect me to rebase (and merge) my changes on current master?

**Comment by rixx** [comment]
> You reorganised the tests last night. Do you expect me to rebase (and merge) my changes on current master?

Nah, that's fine, a rebase would probably be painful. I'll handle it on my end, don't worry about it.

File: src/pretalx/cfp/views/wizard.py [hunk 4]
```diff
@@ -127,6 +129,19 @@ def get_context_data(self, **kwargs):
...
+            for field, model_name in (('submission_type', SubmissionType), ('track', Track)):
+                request_value = self.request.GET.get(field)
+                try:
+                    request_id = model_name.id_from_slug(request_value) if request_value else None
+                except:
+                    continue
+                # search requested object by ID
+                obj = model_name.objects.filter(event=self.request.event, id=request_id).first()
+                # ensure that the whole slug matches, not only the ID
+                if obj and obj.slug() == request_value:
```
Comment (rixx) [inline src/pretalx/cfp/views/wizard.py]: But we can't be sure that this is true: The slug is calculated off `str(obj.name)` after all, which can differ between active locales. I'd just not validate this here.

File: src/pretalx/common/mixins/models.py [hunk 5]
```diff
@@ -38,3 +39,24 @@ def logged_actions(self):
...
+        Get slug of this object.
+        """
+        return f'{self.id}{self.slug_separator}{slugify(self.name)}'
+
+    @classmethod
+    def id_from_slug(cls, slug):
+        """
+        Get ID from slug value.
+        """
+        return int(slug.split(cls.slug_separator)[0])
```
Comment (rixx) [inline src/pretalx/common/mixins/models.py]: This potentially causes a `ValueError` if called with a string that doesn't start with a number.

**Comment by rixx** [comment]
Thank you for this PR! If it's alright by you, I'll address the remaining issues myself.

**Comment by rixx** [comment]
I've rebased and added a couple of fixes and rewrites. Going to fix tests soon, and merge afterwards.

**Pull request merged**